ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/ENRResponsePacketData.java

```diff
@@ -39,24 +46,48 @@ private ENRResponsePacketData(final Bytes requestHash, final NodeRecord enr) {
     this.enr = enr;
   }
 
-  static ENRResponsePacketData create(final Bytes requestHash, final NodeRecord enr) {
+  public static ENRResponsePacketData create(final Bytes requestHash, final NodeRecord enr) {
     return new ENRResponsePacketData(requestHash, enr);
   }
 
   public static ENRResponsePacketData readFrom(final RLPInput in) {
     in.enterList();
     final Bytes requestHash = in.readBytes();
-    final NodeRecord enr =
-        new NodeRecordFactory(IdentitySchemaInterpreter.V4).fromBytes(in.readBytes());
+    in.enterList();
+    final Bytes signature = in.readBytes();
+    final UInt64 sequence = UInt64.fromBytes(RlpUtil.decodeSingleString(in.readBytes()));
+    List<EnrField> enrFields = new ArrayList<>();
+    while (!in.isEndOfCurrentList()) {
+      String key = new String(in.readBytes().toArray(), StandardCharsets.UTF_8);
+      if (key.equals(EnrField.ID)) {
+        enrFields.add(
+            new EnrField(
+                key,
+                IdentitySchema.fromString(
+                    new String(in.readBytes().toArray(), StandardCharsets.UTF_8))));
+      } else if (key.equals(EnrField.TCP)
+          || key.equals(EnrField.UDP)
+          || key.equals(EnrField.TCP_V6)
+          || key.equals(EnrField.UDP_V6)) {
+        enrFields.add(new EnrField(key, in.readIntScalar()));
+      } else {
```

Comment:
This will blow up if we try to read an "eth" element, which is a list (of a list of two numbers).  We will need a clause for `in.nextIsList()` and then handle a (possibly recursive if it's generic) read.

Reply:
Add this to AbstractRLPInput and plumb the method through RLPInput
```
  @Override
  public Bytes currentListAsBytes() {
    if (currentItem >= size) {
      throw error("Cannot read list, input is fully consumed");
    }
    if (!currentKind.isList()) {
      throw error("Cannot read list, current item is not a list list");
    }

    final MutableBytes scratch = MutableBytes.create(currentPayloadSize + 10);
    final int headerSize = RLPEncodingHelpers.writeListHeader(currentPayloadSize, scratch, 0);
    payloadSlice().copyTo(scratch, headerSize);
    final Bytes res = scratch.slice(0, currentPayloadSize + headerSize);

    setTo(nextItem());
    return res;
  }
```

then this whole block becomes

```
    final NodeRecord enr =
        new NodeRecordFactory(IdentitySchemaInterpreter.V4).fromBytes(in.currentListAsBytes());
```

we wind up providing the raw RLP list to the library and it can deal with the nested elements.